algorithms/kernel/dtrees/forest/regression/oneapi/df_regression_train_dense_default_oneapi_impl.i

```diff
@@ -0,0 +1,1245 @@
+    selectParallelizationTechnique<sse2>(par, technique);
+    engines::internal::Params<sse2> params(par.nTrees);
+    for (size_t i = 0; i < par.nTrees; i++)
+    {
+        params.nSkip[i] = i * par.nTrees * x->getNumberOfRows() * (par.featuresPerNode + 1);
+    }
+    DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, par.nTrees, sizeof(engines::EnginePtr));
+    daal::services::internal::TArray<engines::EnginePtr, sse2> engines(par.nTrees);
+    engines::internal::EnginesCollection<sse2> enginesCollection(par.engine, technique, params, engines, &status);
+    DAAL_CHECK_STATUS_VAR(status);
+
+    if (!par.bootstrap)
+    {
+        DAAL_CHECK_STATUS_VAR(initializeTreeOrder(_nSelectedRows, treeOrderLev));
+    }
+
+    for (size_t iter = 0; (iter < par.nTrees) && !algorithms::internal::isCancelled(status, pHostApp); ++iter)
+    {
+        size_t nNodes   = 1; // num of potential nodes to split on current tree level
+        size_t nOOBRows = 0;
+
+        Collection<TreeLevel> DFTreeRecords;
+        Collection<UniversalBuffer> levelNodeLists;    // lists of nodes int props(rowsOffset, rows, ftrId, ftrVal ... )
+        Collection<UniversalBuffer> levelNodeImpLists; // list of nodes fptype props (impurity, mean)
+        UniversalBuffer oobRows;
+
+        levelNodeLists.push_back(context.allocate(TypeIds::id<int>(), nNodes * TreeLevel::_nNodeSplitProps, &status));
+        levelNodeImpLists.push_back(context.allocate(TypeIds::id<algorithmFPType>(), nNodes * TreeLevel::_nNodeImpProps, &status));
+        DAAL_CHECK_STATUS_VAR(status);
+
+        {
+            auto rootNode     = levelNodeLists[0].template get<int>().toHost(ReadWriteMode::writeOnly);
+            rootNode.get()[0] = 0;              // rows offset
+            rootNode.get()[1] = _nSelectedRows; // num of rows
+        }
+
+        auto engineImpl = dynamic_cast<engines::internal::BatchBaseImpl *>(engines[iter].get());
+        if (!engineImpl) return Status(ErrorEngineNotSupported);
+
+        if (par.bootstrap)
+        {
+            // TODO migrate to gpu generators and gpu sort version
+            DAAL_ITTNOTIFY_SCOPED_TASK(compute.RNG);
+            daal::internal::RNGs<int, sse2> rng;
+            rng.uniform(_nSelectedRows, selectedRowsHost.get(), engineImpl->getState(), 0, nRows);
+            daal::algorithms::internal::qSort<int, sse2>(_nSelectedRows, selectedRowsHost.get());
+
+            context.copy(treeOrderLev, 0, (void *)selectedRowsHost.get(), 0, _nSelectedRows, &status);
+            DAAL_CHECK_STATUS_VAR(status);
+        }
+
+        if (oobRequired)
+        {
+            getOOBRows(treeOrderLev, _nSelectedRows, nOOBRows, oobRows); // nOOBRows and oobRows are the output
+        }
+
+        for (size_t level = 0; nNodes > 0; level++)
+        {
+            auto nodeList = levelNodeLists[level];
+            auto impList  = levelNodeImpLists[level];
+
+            daal::services::internal::TArray<int, sse2> selectedFeaturesHost(
```

Comment:
You can use daal_malloc, daal_calloc etc. They dont't have cpu scecifics.

Reply:
yes, it is in my TODO list together with migration to GPU generators and GPU prediction layer